types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,704 @@
+		NULL, drm_lease_request_v1_handle_resource_destroy);
+
+	struct wlr_drm_lease_device_v1 *device =
+		drm_lease_device_v1_from_resource(resource);
+	if (!device) {
+		wlr_log(WLR_ERROR, "Failed to create lease request, "
+				"wlr_drm_lease_device_v1 has been destroyed");
+		return;
+	}
+
+	struct wlr_drm_lease_request_v1 *req =
+		calloc(1, sizeof(struct wlr_drm_lease_request_v1));
+	if (!req) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_request_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Created request %p", req);
+
+	req->device = device;
+	req->resource = request_resource;
+	req->connectors = NULL;
+	req->n_connectors = 0;
+
+	wl_resource_set_user_data(request_resource, req);
+
+	wl_list_insert(&device->requests, &req->link);
+}
+
+static struct wp_drm_lease_device_v1_interface lease_device_impl = {
+	.release = drm_lease_device_v1_handle_release,
+	.create_lease_request = drm_lease_device_v1_handle_create_lease_request,
+};
+
+static void drm_connector_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void drm_connector_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_connector_v1_interface lease_connector_impl = {
+	.destroy = drm_connector_v1_handle_destroy,
+};
+
+static void drm_lease_connector_v1_send_to_client(
+		struct wlr_drm_lease_connector_v1 *connector,
+		struct wl_resource *resource) {
+	if (connector->active_lease) {
+		return;
+	}
+
+	struct wl_client *client = wl_resource_get_client(resource);
+
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *connector_resource = wl_resource_create(client,
+			&wp_drm_lease_connector_v1_interface, version, 0);
+	if (!connector_resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(connector_resource, &lease_connector_impl,
+			connector, drm_connector_v1_handle_resource_destroy);
+	wp_drm_lease_device_v1_send_connector(resource, connector_resource);
+
+	struct wlr_output *output = connector->output;
+	wp_drm_lease_connector_v1_send_name(connector_resource, output->name);
+
+	// TODO: re-send the description when it's updated
+	wp_drm_lease_connector_v1_send_description(connector_resource,
+			output->description);
+
+	wp_drm_lease_connector_v1_send_connector_id(connector_resource,
+			wlr_drm_connector_get_id(output));
+
+	wl_list_insert(&connector->resources,
+			wl_resource_get_link(connector_resource));
+}
+
+static void lease_device_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device = data;
+	if (!device) {
+		wlr_log(WLR_ERROR, "Failed to bind lease device, "
+				"wlr_drm_lease_device_v1 has been destroyed");
+		wl_client_post_no_memory(wl_client);
```

Comment:
It's incorrect to `wl_client_post_no_memory` here. If the device is NULL, it means it's inert and the client is racing with the compositor.

We need to create an inert device resource instead.